controller/grpc.go

```diff
@@ -0,0 +1,1014 @@
+	dataEventListener, err := g.maybeStartEventListener()
+	if err != nil {
+		return nil, protobuf.NewError(err, err.Error())
+	}
+
+	eventListener := &grpcEventListener{
+		Events: make(chan *ct.Event),
+	}
+
+	objectTypeStrings := make([]string, len(objectTypes))
+	for i, t := range objectTypes {
+		objectTypeStrings[i] = string(t)
+	}
+
+	if len(appIDs) == 0 && len(objectIDs) == 0 {
+		// an empty string matches all app ids
+		appIDs = []string{""}
+	}
+	subs := make([]*data.EventSubscriber, 0, len(appIDs)+len(objectIDs))
+	for _, appID := range appIDs {
+		sub, err := dataEventListener.Subscribe(appID, objectTypeStrings, "")
+		if err != nil {
+			return nil, protobuf.NewError(err, err.Error())
+		}
+		subs = append(subs, sub)
+		go (func() {
+			for {
+				event, ok := <-sub.Events
+				if !ok {
+					break
+				}
+				eventListener.Events <- event
+			}
+		})()
+	}
+	for _, objectID := range objectIDs {
+		sub, err := dataEventListener.Subscribe("", objectTypeStrings, objectID)
+		if err != nil {
+			return nil, protobuf.NewError(err, err.Error())
+		}
+		subs = append(subs, sub)
+		go (func() {
+			for {
+				event, ok := <-sub.Events
+				if !ok {
+					break
+				}
+				eventListener.Events <- event
+			}
+		})()
+	}
+	eventListener.subs = subs
+	return eventListener, nil
+}
+
+func (g *grpcAPI) grpcServer() *grpc.Server {
+	s := grpc.NewServer(
+		grpc.StatsHandler(&grpcStatsHandler{}),
+		grpc.StreamInterceptor(g.streamInterceptor),
+		grpc.UnaryInterceptor(g.unaryInterceptor),
+	)
+	protobuf.RegisterControllerServer(s, g)
+	// Register reflection service on gRPC server.
+	reflection.Register(s)
+	return s
+}
+
+type grpcStatsHandler struct{}
+
+func (g *grpcStatsHandler) TagRPC(ctx context.Context, rpcTagInfo *stats.RPCTagInfo) context.Context {
+	return ctx
+}
+
+func (g *grpcStatsHandler) HandleRPC(context.Context, stats.RPCStats) {
+}
+
+const ctxKeyRemoteHost = "remote-host"
+
+func (g *grpcStatsHandler) TagConn(ctx context.Context, connTagInfo *stats.ConnTagInfo) context.Context {
+	remoteHost, _, _ := net.SplitHostPort(connTagInfo.RemoteAddr.String())
+	ctx = context.WithValue(ctx, ctxKeyRemoteHost, remoteHost)
+	return ctx
+}
+
+func (g *grpcStatsHandler) HandleConn(context.Context, stats.ConnStats) {
+}
+
+func (g *grpcAPI) streamInterceptor(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) (err error) {
+	ctx, logRequestEnd := g.logRequest(stream.Context(), info.FullMethod)
+	defer func() {
+		logRequestEnd(ctx, err)
+	}()
+
+	ctx, err = g.authorize(stream.Context())
+	if err != nil {
+		return err
+	}
+
+	wrappedStream := middleware.WrapServerStream(stream)
+	wrappedStream.WrappedContext = ctx
+	return handler(srv, wrappedStream)
+}
+
+func (g *grpcAPI) unaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (res interface{}, err error) {
+	var logRequestEnd func(context.Context, error)
```

Comment:
Is this declaration needed?

Reply:
Seems it could be dropped if the following line uses `:=`.